src/server/service/kv.rs

```diff
@@ -119,32 +120,50 @@ impl<T: RaftStoreRouter + 'static, E: Engine, L: LockManager> Service<T, E, L> {
 
 impl<T: RaftStoreRouter + 'static, E: Engine, L: LockManager> Tikv for Service<T, E, L> {
     fn kv_get(&mut self, ctx: RpcContext<'_>, req: GetRequest, sink: UnarySink<GetResponse>) {
-        let timer = GRPC_MSG_HISTOGRAM_VEC.kv_get.start_coarse_timer();
+        let instant = Instant::now_coarse();
         let future = future_get(&self.storage, req)
             .and_then(|res| sink.success(res).map_err(Error::from))
-            .map(|_| timer.observe_duration())
+            .map(move |_| {
+                GRPC_MSG_HISTOGRAM.may_flush(|m| {
+                    m.kv_get.observe(duration_to_sec(instant.elapsed()));
+                })
+            })
             .map_err(move |e| {
                 debug!("kv rpc failed";
                     "request" => "kv_get",
                     "err" => ?e
                 );
-                GRPC_MSG_FAIL_COUNTER.kv_get.inc();
+                GRPC_MSG_HISTOGRAM.may_flush(|m| {
```

Comment:
@BusyJay Previously looks like duration is recorded only when the request is success. Is it correct?

Reply:
@breeswish As my understand of the code,I think duration is actually recorded no matter request is success or fail. If the original intention is record duration when success,previous implementation might be wrong.